flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java

```diff
@@ -16,15 +16,253 @@
 
 /** Serializer for converting Avro's {@link GenericRecord} to BigQuery proto. */
 public class AvroToProtoSerializer implements BigQueryProtoSerializer<GenericRecord> {
 
+    private Descriptor descriptor;
+
+    /**
+     * Prepares the serializer before its serialize method can be called. It allows contextual
+     * preprocessing after constructor and before serialize. The Sink will internally call this
+     * method when initializing itself.
+     *
+     * @param bigQuerySchemaProvider {@link BigQuerySchemaProvider} for the destination table.
+     */
+    @Override
+    public void init(BigQuerySchemaProvider bigQuerySchemaProvider) {
+        Preconditions.checkNotNull(
+                bigQuerySchemaProvider,
+                "BigQuerySchemaProvider not found while initializing AvroToProtoSerializer");
+        Descriptor derivedDescriptor = bigQuerySchemaProvider.getDescriptor();
+        Preconditions.checkNotNull(
+                derivedDescriptor, "Destination BigQuery table's Proto Schema could not be found.");
+        this.descriptor = derivedDescriptor;
+    }
+
     @Override
     public ByteString serialize(GenericRecord record) throws BigQuerySerializationException {
-        throw new UnsupportedOperationException("serialize method is not supported");
+        try {
+            return getDynamicMessageFromGenericRecord(record, this.descriptor).toByteString();
+        } catch (Exception e) {
+            throw new BigQuerySerializationException(e.getMessage(), e);
+        }
+    }
+
+    /**
+     * Function to convert a Generic Avro Record to Dynamic Message to write using the Storage Write
+     * API.
+     *
+     * @param element {@link GenericRecord} Object to convert to {@link DynamicMessage}
+     * @param descriptor {@link Descriptor} describing the schema of the sink table.
+     * @return {@link DynamicMessage} Object converted from the Generic Avro Record.
+     */
+    public static DynamicMessage getDynamicMessageFromGenericRecord(
+            GenericRecord element, Descriptor descriptor) {
+        Schema recordSchema = element.getSchema();
+        DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+        // Get the record's schema and find the field descriptor for each field one by one.
+        for (Schema.Field field : recordSchema.getFields()) {
+            // In case no field descriptor exists for the field, throw an error as we have
+            // incompatible schemas.
+            FieldDescriptor fieldDescriptor =
+                    Preconditions.checkNotNull(
+                            descriptor.findFieldByName(field.name().toLowerCase()));
+            // Get the value for a field.
+            // Check if the value is null.
+            @Nullable Object value = element.get(field.name());
+            if (value == null) {
+                // If the field required, throw an error.
+                if (fieldDescriptor.isRequired()) {
```

Comment:
For the cases where value == null but the fieldDescriptor is `REPEATED` or `NULLABLE` we do nothing (this fieldDescriptor is absent in the `DynamicMessage`, Storage API automatically adds it as `NULL` or `[]` in case of `REPEATED` fields).

Reply:
that makes sense. that behavior would remain intact if you consolidate the nested if statement into one-line, right?

Reply:
Removing the nesting would cause the case of `value == null` and not `isRequired()` to fall under the else case which is not desired.

Reply:
okay you can resolve this